### ISSUE TITLE
Rule proposal: jsx-tag-spacing

### ISSUE THREAD

**Issue by Kovensky** [yannickcr/eslint-plugin-react#693]
For some reason, `/ >` is valid JSX, as is `/` followed by any kind of whitespace until the `>`. This rule would require that the self-closing terminator is always written as an atomic `/>`.

Not having a check for this also allows for funny workarounds for [jsx-closing-bracket-location](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-closing-bracket-location.md), such as this (when using `after-props`):

``` js
<SomeComponent
  with='many'
  props='specified' /
>
```

Update: https://github.com/yannickcr/eslint-plugin-react/issues/693#issuecomment-234842584

**Comment by ljharb** [comment]
I'm a strong +1 on this - there should never be a space between `/` and `>` and I'm horrified that someone would ever come up with that.

**Comment by Kovensky** [comment]
It also turns out that `<Tag><   /Tag>` is weird but valid JSX too, so I made the rule also check for this.

**Comment by ljharb** [comment]
Also `</ Tag>` ideally.

**Comment by Kovensky** [comment]
I can implement that, but I think it should be a separate rule. This one concerns itself with space in the opener/closer itself, while the other rule would worry about whitespace between the opener and the component name (analogous to jsx-space-before-closing).

**Comment by Kovensky** [comment]
Per discussion with @ljharb on the babel slack:

Merge this proposed rule with `jsx-space-before-closing` and a rule for checking space after opening, with a default schema like:

```js
{
  closingSlash: 'never', // internal space between the "</" and "/>" tokens, e.g. "< /" would warn
  beforeClosing: 'always', // same thing as the current jsx-space-before-closing option
  afterOpening: 'never', // same thing as jsx-space-before-closing, but right after the "<" or "</"
}
```

`afterOpening` would also have an `'allow-multiline'` suboption. Adding an `'allow-multiline'` suboption to `beforeClosing` (and using it as default instead) is possible, but a stricter `beforeClosing: 'always'` would conflict with `jsx-closing-bracket-location`.

**Comment by ljharb** [comment]
Let's also rename this rule to `jsx-tag-spacing` - and then we can deprecate `jsx-closing-bracket-location` and copy the tests over.

**Comment by Kovensky** [comment]
Because `jsx-space-before-closing` only works on self-closing tags, I'll name the equivalent option `beforeSelfClosing`, and have `beforeClosing` apply to regular closing tags instead. The defaults would be `beforeSelfClosing: 'always'` and `beforeClosing: 'never'`.
